allennlp/data/iterators/cached_iterator.py

```diff
@@ -0,0 +1,45 @@
+import logging
+import random
+from typing import Iterable, Dict, List
+from overrides import overrides
+from allennlp.data.instance import Instance
+from allennlp.data.iterators.bucket_iterator import BucketIterator
+from allennlp.data.iterators.data_iterator import DataIterator
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@DataIterator.register("cached")
+class CachedIterator(BucketIterator):
+    """
+    An extension of the `BucketIterator` that caches its batches between epochs. Batches are built
+    only once during the first epoch, then the same batches are used again in consecutive epochs.
+    Using this iterator saves run time but costs more memory to store the cached batches. It should
+    be used only if the dataset can fit in memory. This iterator shouldn't be used with the lazy
+    mode.
+
+    Parameters:
+    ----------
+        See :class:`BucketIterator`.
+    """
```

Comment:
I would implement a constructor that throws a `ConfigurationError` if you try to instantiate it with lazy=True

Reply:
although actually, laziness is a function of the dataset, not of the iterator, so this doesn't make sense